File: web/src/next/panel/alignment.ts
/**
 * The alignment finder on the Sun and Moon cards (expansion programme Q8, photo agent):
 * "When does the Sun set (or rise) along this line?" — down a street (Manhattanhenge), through
 * a window, over a landmark. The bearing is typed, or picked on the map (a click there sets
 * the bearing from the place, and the ray is drawn); the event is setting, rising, or the
 * centre standing at a chosen height (a skyline); a tolerance; a year. The answer lists the
 * days, the closest of each run marked, and pressing a day shows it.
 *
 * The search is the engine's `alignment_days` (suntools, CONVENTIONS 13.10): a whole year in
 * one call, about 0.2 s for the Sun and 0.9 s for the Moon in WebAssembly, so it runs only
 * when the person presses Find, never on its own. The rows are a pure function
 * (`alignmentRows`), tested with the mock engine in photo-tools.test.ts.
 */

import { h } from '../../dom.js';
import type { Ctx } from '../component.js';
import {
  isGeomagEngine,
  isSunToolsEngine,
  type AlignmentEvent,
  type AlignmentMatch,
  type AlignmentRequest,
  type AlignmentResult,
  type Observer,
} from '../engine/types.js';
import { setTime } from '../playback.js';
import { setText } from '../shell/derived.js';
import { dateShort, eventTime, formatAngle } from '../shell/format.js';
import { displayZone, engineObserver, type AngleFormat, type ExplorerState } from '../state.js';
import { icon } from '../theme/icons.js';
import { segmented } from '../theme/primitives.js';
import { msFromJd, wallClock, type Zone } from '../time.js';
import { formatYear, gregorianDateOfMs, turning } from '../time/index.js';
import { bearingSourceText, deltaTNote, photoBearing, toolChip, type BearingValue } from './photo.js';
import { magneticFromTrue, parseAltitude, parseBearing, parseTolerance, TOLERANCE_MAX, TOLERANCE_MIN } from './sun-tools.js';

export type AlignmentEventKind = 'set' | 'rise' | 'at_altitude';

/** What the person chose, kept per explorer page (never stored). */
interface Choices {
  event: AlignmentEventKind;
  height: string;
  tolerance: string;
  /** Years from the year of the time shown. */
  yearOffset: number;
}

const choices = new WeakMap<object, Choices>();

function choicesFor(store: object): Choices {
  let c = choices.get(store);
  if (!c) {
    c = { event: 'set', height: '0.5', tolerance: '0.5', yearOffset: 0 };
    choices.set(store, c);
  }
  return c;
}

export interface AlignmentRow {
  jd: number;
  /** `Tue 26 May`, on the display clock. */
  date: string;
  /** `20:18`. */
  time: string;
  /** `sets`, `rises`, `sinks through 0.5°`. */
  what: string;
  /** `299.0°`. */
  az: string;
  /** `0.03° right of the line`. */
  offset: string;
  /** `0.03° right`, for the list. */
  offsetShort: string;
  best: boolean;
  /** For assistive technology and the button's name. */
  label: string;
}

/** Offsets smaller than this read as "on the line", degrees. */
const ON_LINE_DEG = 0.005;

/** Where the body is against the line, looking along the bearing: clockwise is to the right. */
export function offsetWords(offsetDeg: number): string {
  if (Math.abs(offsetDeg) < ON_LINE_DEG) return 'on the line';
  const a = Math.abs(offsetDeg);
  return `${a < 1 ? a.toFixed(2) : a.toFixed(1)}° ${offsetDeg > 0 ? 'right' : 'left'} of the line`;
}

function kindWords(m: Pick<AlignmentMatch, 'kind'>, event: AlignmentEvent, format: AngleFormat): string {
  switch (m.kind) {
    case 'rise':
      return 'rises';
    case 'set':
      return 'sets';
    case 'rising':
    case 'setting': {
      const hText = event.kind === 'at_altitude' ? formatAngle(event.altitude_deg, format, 'coarse') : '';
      return `${m.kind === 'rising' ? 'climbs' : 'sinks'} through ${hText}`;
    }
  }
}

function eventName(event: AlignmentEvent, format: AngleFormat): string {
  return event.kind === 'set'
    ? 'sets'
    : event.kind === 'rise'
      ? 'rises'
      : `stands ${formatAngle(event.altitude_deg, format, 'coarse')} high`;
}

/** Runs of consecutive matching days: the engine marks the closest of each `best`. */
function runCount(matches: readonly AlignmentMatch[]): number {
  return matches.filter((m) => m.best).length;
}

/**
 * The list of days and one sentence for an `alignment_days` result. `dt` is the ± text after a
 * time (` ±6 s`): one for every row, or worked out per row (chip2: the Moon's share of σ(ΔT)
 * changes through the month).
 */
export function alignmentRows(
  result: AlignmentResult,
  zone: Zone,
  format: AngleFormat,
  dt: string | ((jd: number) => string) = '',
): { rows: AlignmentRow[]; summary: string } {
  const note = typeof dt === 'function' ? dt : () => dt;
  const body = result.body;
  const along = `${result.azimuth_deg.toFixed(1)}°`;
  const within = `within ${result.tolerance_deg}° of ${along}`;
  const rows = result.matches.map((m): AlignmentRow => {
    const date = dateShort(m.jd_utc, zone);
    const time = `${eventTime(m.jd_utc, zone)}${note(m.jd_utc)}`;
    const what = kindWords(m, result.event, format);
    const offset = offsetWords(m.offset_deg);
    return {
      jd: m.jd_utc,
      date,
      time,
      what,
      az: `${m.az_deg.toFixed(1)}°`,
      offset,
      offsetShort: offset.replace(/ of the line$/, ''),
      best: m.best,
      label: `Show ${date} ${time}: the ${body} ${what} at ${m.az_deg.toFixed(1)}°, ${offset}${m.best ? ', the closest day of its run' : ''}`,
    };
  });
  const cut = result.truncated ? ' Part of the year is outside the years the core covers.' : '';
  let summary: string;
  if (rows.length) {
    const runs = runCount(result.matches);
    summary = `The ${body} ${eventName(result.event, format)} ${within} on ${rows.length} ${rows.length === 1 ? 'day' : 'days'} of ${result.year}${runs > 1 ? `, in ${runs} runs; the closest day of each is marked` : rows.length > 1 ? '; the closest is marked' : ''}. Press a day to show it.${cut}`;
  } else if (result.closest) {
    const c = result.closest;
    summary = `The ${body} never ${eventName(result.event, format)} ${within} in ${result.year}. The nearest is ${dateShort(c.jd_utc, zone)} at ${eventTime(c.jd_utc, zone)}${note(c.jd_utc)}, ${offsetWords(c.offset_deg)}: widen the tolerance, or check the bearing.${cut}`;
  } else {
    summary = `The ${body} never ${result.event.kind === 'set' ? 'sets' : result.event.kind === 'rise' ? 'rises' : 'reaches that height'} here in ${result.year}.${cut}`;
  }
  return { rows, summary };
}

export interface AlignmentTool {
  el: HTMLElement;
  /** On each render of the Selected card (`body` is the Sun or the Moon). */
  update(s: ExplorerState, body: string): void;
  destroy(): void;
}

let seq = 0;

export function alignmentTool(ctx: Ctx): AlignmentTool {
  const { store, engine } = ctx;
  const id = `sf-align-${++seq}`;
  const c = choicesFor(store);
  const bearing = photoBearing(ctx);
  const summaryText = h('span', {}, 'Sunset or sunrise along a line');
  const intro = h('p', { class: 'sf-photo__intro' });

  // --- the bearing: typed, or picked on the map -----------------------------------------
  const bearingInput = h('input', {
    class: 'sf-input sf-num sf-photo__input',
    id: `${id}-az`,
    type: 'text',
    inputmode: 'decimal',
    autocomplete: 'off',
    spellcheck: 'false',
    placeholder: '299',
    value: bearing.get().text,
    'aria-describedby': `${id}-az-note`,
  });
  const pickBtn = h(
    'button',
    { type: 'button', class: 'sf-btn sf-btn--secondary sf-btn--sm sf-photo__pick', 'data-tip': 'Click a point on the map: the bearing becomes the direction from your place to it' },
    icon('target'),
    h('span', { class: 'sf-btn__label' }, 'Pick on the map'),
  );
  const bearingNote = h('p', { class: 'sf-photo__sub', id: `${id}-az-note` });

  // --- the event, its height, the tolerance and the year ------------------------------------
  const eventSeg = segmented<AlignmentEventKind>({
    label: 'Which moment',
    size: 'sm',
    value: c.event,
    options: [
      { value: 'set', label: 'Sets', tip: 'Its upper edge touches a sea-level horizon, as the rise and set times give it' },
      { value: 'rise', label: 'Rises', tip: 'Its upper edge touches a sea-level horizon, as the rise and set times give it' },
      { value: 'at_altitude', label: 'At a height', tip: 'Its centre at a height you choose, refraction included: for a skyline or a hill' },
    ],
    onChange: (v) => {
      c.event = v;
      syncHeight();
      markChanged();
    },
  });
  const heightInput = h('input', {
    class: 'sf-input sf-num sf-photo__input',
    id: `${id}-h`,
    type: 'text',
    inputmode: 'decimal',
    autocomplete: 'off',
    spellcheck: 'false',
    value: c.height,
  });
  const heightRow = h(
    'div',
    { class: 'sf-photo__ask' },
    h('label', { class: 'sf-photo__label', for: `${id}-h`, 'data-tip': 'The height of its centre as you would see it, refraction included. 0.5° puts the centre on a sea horizon without refraction (the “half Sun” of the Manhattanhenge dates newspapers print).' }, 'Height of its centre'),
    heightInput,
    h('span', { class: 'sf-photo__unit' }, '°'),
  );
  const tolInput = h('input', {
    class: 'sf-input sf-num sf-photo__input',
    id: `${id}-tol`,
    type: 'text',
    inputmode: 'decimal',
    autocomplete: 'off',
    spellcheck: 'false',
    value: c.tolerance,
  });
  const yearValue = h('span', { class: 'sf-num sf-photo__year' });
  const prevYear = h('button', { type: 'button', class: 'sf-btn sf-btn--ghost sf-btn--sm sf-btn--icon', 'aria-label': 'The year before' }, icon('chevron-left'));
  const nextYear = h('button', { type: 'button', class: 'sf-btn sf-btn--ghost sf-btn--sm sf-btn--icon', 'aria-label': 'The year after' }, icon('chevron-right'));
  const find = h('button', { type: 'button', class: 'sf-btn sf-btn--primary sf-btn--sm sf-photo__find' }, h('span', { class: 'sf-btn__label' }, 'Find the days'));
  const statusText = h('span', {});
  const chip = toolChip();
  const status = h('p', { class: 'sf-photo__status', role: 'status', 'aria-live': 'polite' }, statusText, ' ', chip.el);
  const list = h('ul', { class: 'sf-photo__list', 'aria-label': 'Days on the line' });
  // The Sun's rising and setting bearings through the year (charts2's Sun tab), on demand.
  const bearingsChart = h('button', { type: 'button', class: 'sf-link sf-photo__chartlink' }, 'The Sun’s bearings through the year (Charts)');
  bearingsChart.addEventListener('click', () => {
    void import('../charts/index.js').then((m) => m.showCharts(store, 'sun', 'bearings'));
  });

  const el = h(
    'details',
    { class: 'sf-details sf-photo-align' },
    h('summary', {}, summaryText, icon('chevron-down')),
    h(
      'div',
      { class: 'sf-photo__body' },
      intro,
      h(
        'div',
        { class: 'sf-photo__ask' },
        h('label', { class: 'sf-photo__label', for: `${id}-az` }, 'Bearing from here'),
        bearingInput,
        h('span', { class: 'sf-photo__unit' }, '° true'),
      ),
      h('div', { class: 'sf-photo__pickrow' }, pickBtn),
      bearingNote,
      h('div', { class: 'sf-photo__ask' }, h('span', { class: 'sf-photo__label' }, 'Moment'), eventSeg.el),
      heightRow,
      h(
        'div',
        { class: 'sf-photo__ask' },
        h('label', { class: 'sf-photo__label', for: `${id}-tol`, 'data-tip': 'How far from the line still counts, either side' }, 'Within'),
        tolInput,
        h('span', { class: 'sf-photo__unit' }, '° either side'),
      ),
      h('div', { class: 'sf-photo__ask' }, h('span', { class: 'sf-photo__label' }, 'Year'), prevYear, yearValue, nextYear),
      find,
      status,
      list,
      bearingsChart,
    ),
  ) as HTMLDetailsElement;

  let last: { s: ExplorerState; body: string } | null = null;
  /** What the shown list answers, to tell when the inputs changed since. */
  let shownKey = '';
  let releaseRay: (() => void) | null = null;
  let busy = false;

  const syncHeight = (): void => {
    heightRow.hidden = c.event !== 'at_altitude';
  };
  syncHeight();

  /** The year the engine lays out (the wire's proleptic Gregorian year of the local time). */
  const yearOf = (s: ExplorerState): number => {
    const w = wallClock(s.time.jd_utc, displayZone(s));
    return gregorianDateOfMs(msFromJd(s.time.jd_utc) + w.offsetMs).year + c.yearOffset;
  };

  /** The request as the inputs stand, or a sentence saying what is wrong. */
  const request = (
    s: ExplorerState,
    body: string,
  ): { req: AlignmentRequest; obs: Observer; key: string } | string => {
    const az = bearing.get().azimuth;
    if (az === null) return 'Type a bearing (degrees from true north, 0 to 360, or a compass point such as WNW), or pick one on the map.';
    const tol = parseTolerance(tolInput.value);
    if (tol === null) return `Type a tolerance from ${TOLERANCE_MIN}° to ${TOLERANCE_MAX}°.`;
    let event: AlignmentEvent;
    if (c.event === 'at_altitude') {
      const hDeg = parseAltitude(heightInput.value);
      if (hDeg === null || hDeg < -5) return 'Type the height of its centre, from −5° to 90°.';
      event = { kind: 'at_altitude', altitude_deg: hDeg };
    } else {
      event = { kind: c.event };
    }
    const zone = displayZone(s);
    const offset = wallClock(s.time.jd_utc, zone).offsetMs / 3_600_000;
    const req: AlignmentRequest = { body, year: yearOf(s), azimuth_deg: az, tolerance_deg: tol, event, utc_offset_hours: offset };
    const obs = engineObserver(s);
    return { req, obs, key: JSON.stringify([req, obs]) };
  };
  const sunEngine = isSunToolsEngine(engine) ? engine : null;

  const describe = (s: ExplorerState, body: string): void => {
    setText(summaryText, body === 'Moon' ? 'Moonrise or moonset along a line' : 'Sunrise or sunset along a line');
    bearingsChart.hidden = body !== 'Sun';
    setText(
      intro,
      body === 'Moon'
        ? 'The days of a year when the Moon rises or sets along a line: over a landmark, down a valley. Type the bearing or pick it on the map.'
        : 'The days of a year when the Sun sets or rises along a line: down a street (Manhattanhenge), through a window, over a landmark. Type the bearing or pick it on the map.',
    );
    setText(yearValue, formatYear(yearOf(s)));
    const v = bearing.get();
    if (document.activeElement !== bearingInput && bearingInput.value !== v.text) bearingInput.value = v.text;
    bearingInput.toggleAttribute('aria-invalid', bearingInput.value.trim() !== '' && v.azimuth === null);
    setText(bearingNote, bearingNoteText(v, s, ctx));
    pickBtn.setAttribute('aria-pressed', String(bearing.picking()));
    setText(pickBtn.querySelector('.sf-btn__label')!, bearing.picking() ? 'Click the map… (Esc stops)' : 'Pick on the map');
  };

  const markChanged = (): void => {
    if (!last || busy) return;
    const r = request(last.s, last.body);
    const stale = typeof r === 'string' || r.key !== shownKey;
    list.toggleAttribute('data-stale', stale && list.childElementCount > 0);
    if (stale && list.childElementCount > 0) setText(statusText, 'The line, the moment or the place changed: press Find the days again.');
  };

  const run = (): void => {
    if (busy) return;
    // Before the card's first frame (a hidden tab draws none), from the store as it is.
    const s = last?.s ?? store.get();
    const body = last?.body ?? (s.selection.body === 'Moon' ? 'Moon' : 'Sun');
    if (!sunEngine) {
      setText(statusText, 'The alignment finder is not available in this engine: rebuild the WebAssembly package.');
      return;
    }
    const r = request(s, body);
    if (typeof r === 'string') {
      setText(statusText, r);
      list.replaceChildren();
      return;
    }
    busy = true;
    find.setAttribute('aria-busy', 'true');
    find.disabled = true;
    setText(statusText, `Searching ${r.req.year}${body === 'Moon' ? ' (about a second for the Moon)' : ''}…`);
    // Let the words paint before the search takes the page's time.
    requestAnimationFrame(() =>
      setTimeout(() => {
        const st = store.get();
        let result: AlignmentResult | null = null;
        let error = '';
        try {
          result = sunEngine.alignmentDays(r.obs, r.req);
        } catch (e) {
          error = e instanceof Error ? e.message : String(e);
        }
        busy = false;
        find.removeAttribute('aria-busy');
        find.disabled = false;
        list.removeAttribute('data-stale');
        if (!result) {
          list.replaceChildren();
          setText(statusText, `Not found: ${error}.`);
          return;
        }
        shownKey = r.key;
        const zone = displayZone(st);
        // A rising, setting or height on a line: set by the Earth's turning, the body's own
        // share of σ(ΔT) (chip2).
        chip.set(ctx, turning(result.body), result.jd_start, result.jd_end);
        const model = alignmentRows(result, zone, st.settings.angleFormat, (jd) => deltaTNote(ctx, jd, turning(result.body)));
        setText(statusText, model.summary);
        list.replaceChildren(
          ...model.rows.map((row) => {
            const b = h(
              'button',
              { type: 'button', class: `sf-photo__row${row.best ? ' sf-photo__row--best' : ''}`, 'aria-label': row.label },
              h('span', { class: 'sf-photo__date' }, row.date),
              h('span', { class: 'sf-num' }, row.time),
              // "sets" is in the sentence above; a height's crossing says which way.
              h(
                'span',
                { class: 'sf-photo__what' },
                /^(climbs|sinks)/.test(row.what) ? `${row.what.split(' ')[0]}, ` : '',
                h('span', { class: 'sf-num' }, row.az),
                `, ${row.offsetShort}`,
              ),
              row.best ? h('span', { class: 'sf-photo__badge' }, 'best') : null,
            );
            b.addEventListener('click', () => setTime(store, row.jd));
            return h('li', {}, b);
          }),
        );
      }, 0),
    );
  };

  // --- wiring ------------------------------------------------------------------------------
  bearingInput.addEventListener('input', () => {
    bearing.setTyped(bearingInput.value, parseBearing(bearingInput.value));
  });
  for (const input of [bearingInput, heightInput, tolInput]) {
    input.addEventListener('keydown', (e) => {
      if (e.key === 'Enter') {
        e.preventDefault();
        run();
      }
    });
  }
  heightInput.addEventListener('input', () => {
    c.height = heightInput.value;
    heightInput.toggleAttribute('aria-invalid', parseAltitude(heightInput.value) === null);
    markChanged();
  });
  tolInput.addEventListener('input', () => {
    c.tolerance = tolInput.value;
    tolInput.toggleAttribute('aria-invalid', parseTolerance(tolInput.value) === null);
    markChanged();
  });
  const stepYear = (n: number): void => {
    c.yearOffset += n;
    if (last) setText(yearValue, formatYear(yearOf(last.s)));
    markChanged();
  };
  prevYear.addEventListener('click', () => stepYear(-1));
  nextYear.addEventListener('click', () => stepYear(1));
  find.addEventListener('click', run);
  pickBtn.addEventListener('click', () => {
    if (bearing.picking()) {
      bearing.cancelPick();
      return;
    }
    const s = store.get();
    // The picking happens on the map: show it if another view is up.
    if (s.view !== 'map' && s.view !== 'globe') store.patch({ view: 'map' });
    bearing.pick({ lat_deg: s.observer.lat_deg, lon_deg: s.observer.lon_deg }, () => bearingInput.focus({ preventScroll: true }));
  });
  const stopBearing = bearing.subscribe(() => {
    if (last) describe(last.s, last.body);
    markChanged();
  });
  const syncRay = (): void => {
    if (el.open && !releaseRay) releaseRay = bearing.holdRay();
    else if (!el.open && releaseRay) {
      releaseRay();
      releaseRay = null;
    }
  };
  el.addEventListener('toggle', syncRay);

  return {
    el,
    update(s, body) {
      const prev = last;
      last = { s, body };
      if (!prev || prev.body !== body || prev.s.observer !== s.observer || prev.s.settings !== s.settings || yearOf(prev.s) !== yearOf(s)) {
        describe(s, body);
        markChanged();
      }
    },
    destroy() {
      stopBearing();
      releaseRay?.();
      releaseRay = null;
      if (bearing.picking()) bearing.cancelPick();
    },
  };
}

/** Under the bearing: where it came from, and the same bearing on a magnetic compass. */
function bearingNoteText(v: BearingValue, s: ExplorerState, ctx: Ctx): string {
  if (v.azimuth === null) return 'Degrees from true north, clockwise (90 east, 270 west), or a compass point: WNW.';
  const here = { lat_deg: s.observer.lat_deg, lon_deg: s.observer.lon_deg };
  const source = bearingSourceText(v, here);
  let magnetic = '';
  const engine = ctx.engine;
  if (isGeomagEngine(engine)) {
    try {
      const f = engine.magneticField(here.lat_deg, here.lon_deg, s.observer.height_m, Math.floor(s.time.jd_utc));
      if (f.available) magnetic = ` On a compass: ${magneticFromTrue(v.azimuth, f.declination_deg).toFixed(1)}° magnetic (variation ${f.variation_text}).`;
    } catch {
      magnetic = '';
    }
  }
  return `${source ? `${source} ` : ''}The line is drawn on the map.${magnetic}`;
}

File: crates/skyfix-almanac/src/sun_tools/alignment.rs
//! The alignment finder: on which days of a year does a body rise or set along a given
//! bearing, or stand at a given height on it (CONVENTIONS 13.10)?
//!
//! Manhattanhenge is the example everyone knows: Manhattan's street grid runs 29° off
//! true east-west, so the Sun sets straight down the cross streets on two evenings in
//! late May and two in mid-July. The same question fits a window, a stone row, a street
//! or a valley for the Sun, the Moon, a planet or a star.
//!
//! - `rise` / `set`: the event finder's rise and set (`day_events`, CONVENTIONS 13.3): the
//!   instant the **upper limb** touches the sea-level horizon with standard refraction
//!   (the centre at `h0`: -50' for the Sun, `-34' - SD` for the Moon, -34' otherwise;
//!   lowered by the dip with `horizon = dip`). The azimuth reported is the centre's at
//!   that instant.
//! - `at_altitude`: the instants the centre's **apparent** altitude is `altitude_deg`
//!   (`find_altitude`), rising and setting. `altitude_deg = 0` is the photographers'
//!   "half sun" on a flat horizon, and `+SD` (about 0.27° for the Sun) the "full sun"
//!   sitting on it.
//!
//! The whole year is one search (one `day_events` or one `find_altitude` over the local
//! year, at most 366 days), so the events are exactly those the explorer shows. A day
//! matches when its event's azimuth is within `tolerance_deg` of the bearing; each run of
//! consecutive matching days marks its closest day `best`. `closest` is the year's
//! closest event whether it matches or not, so "never" can say by how much.

use serde::{Deserialize, Serialize};
use skyfix_core::time::format_utc;
use skyfix_core::units::norm_180;
use skyfix_ephemeris::body::BodyEphemeris;
use skyfix_ephemeris::topocentric::Site;

use super::{clip_to_coverage, clock_offset_hours, local_date, year_window};
use crate::events::{EventKind, EventOptions, day_events, find_altitude};
use crate::sky::{AlmanacError, checked_site, resolve_bodies};

/// Which moment of the day is aligned.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum AlignmentEvent {
    Rise,
    Set,
    /// The centre at this apparent altitude, rising and setting.
    AtAltitude {
        altitude_deg: f64,
    },
}

/// `alignment_days` request (EXPLORER_API.md).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AlignmentRequest {
    #[serde(default = "default_body")]
    pub body: String,
    pub year: i32,
    /// The bearing, degrees from true north.
    pub azimuth_deg: f64,
    /// How far from the bearing still counts, degrees (default 0.5).
    #[serde(default = "default_tolerance")]
    pub tolerance_deg: f64,
    pub event: AlignmentEvent,
    /// The clock the dates are written on; default local mean time.
    #[serde(default)]
    pub utc_offset_hours: Option<f64>,
    /// Rise and set only: the horizon (CONVENTIONS 13.3).
    #[serde(default)]
    pub options: Option<EventOptions>,
}

fn default_body() -> String {
    "Sun".to_string()
}

fn default_tolerance() -> f64 {
    0.5
}

/// Direction of an aligned moment.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AlignedKind {
    Rise,
    Set,
    /// `at_altitude`, climbing through the altitude.
    Rising,
    /// `at_altitude`, sinking through it.
    Setting,
}

/// One aligned (or, for `closest`, nearest) moment.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AlignmentMatch {
    /// Local calendar date on the request's clock, `YYYY-MM-DD`.
    pub date: String,
    pub kind: AlignedKind,
    pub jd_utc: f64,
    pub utc: String,
    /// The centre's azimuth at that instant.
    pub az_deg: f64,
    /// `az_deg - azimuth_deg`, wrapped into (-180, 180].
    pub offset_deg: f64,
    /// Topocentric geometric altitude of the centre at that instant.
    pub alt_deg: f64,
    /// The closest day of its run of consecutive matching days.
    pub best: bool,
}

/// `alignment_days` result.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AlignmentResult {
    pub body: String,
    pub year: i32,
    pub azimuth_deg: f64,
    pub tolerance_deg: f64,
    pub event: AlignmentEvent,
    /// The clock the dates are on (local mean time when the request gave none).
    pub utc_offset_hours: f64,
    /// The window searched: the local year, clipped to the coverage.
    pub jd_start: f64,
    pub jd_end: f64,
    /// The local year reaches outside the coverage and was clipped.
    pub truncated: bool,
    /// Events of the requested kind found in the year.
    pub events_considered: usize,
    /// Every event within the tolerance, time-ordered.
    pub matches: Vec<AlignmentMatch>,
    /// The year's event nearest the bearing, matching or not; `None` when the body has
    /// no such event all year (always up, always down, never at that altitude).
    pub closest: Option<AlignmentMatch>,
}

/// Days of `year` on which `request.body`'s rise, set or given altitude falls within
/// `tolerance_deg` of `azimuth_deg` (EXPLORER_API.md `alignment_days`).
pub fn alignment_days(
    eph: &dyn BodyEphemeris,
    site: &Site,
    request: &AlignmentRequest,
) -> Result<AlignmentResult, AlmanacError> {
    let site = checked_site(site)?;
    let body = resolve_bodies(&[request.body.as_str()])?[0];
    let bearing = request.azimuth_deg;
    if !(bearing.is_finite() && (-360.0..=720.0).contains(&bearing)) {
        return Err(AlmanacError::invalid(format!(
            "azimuth_deg must be a bearing in degrees, got {bearing}"
        )));
    }
    let tol = request.tolerance_deg;
    if !(tol.is_finite() && tol > 0.0 && tol <= 90.0) {
        return Err(AlmanacError::invalid(format!(
            "tolerance_deg must be above 0 and at most 90, got {tol}"
        )));
    }
    let offset = clock_offset_hours(request.utc_offset_hours, &site)?;
    let (y0, y1) = year_window(request.year, offset)?;
    let w = clip_to_coverage(eph, body, y0, y1)?;

    // (kind, jd, az, alt) for every event of the requested sort in the year.
    let mut events: Vec<(AlignedKind, f64, f64, f64)> = Vec::new();
    match request.event {
        AlignmentEvent::Rise | AlignmentEvent::Set => {
            let options = request.options.unwrap_or_default();
            let de = day_events(eph, &site, w.start, w.end, &[body], &options)?;
            if let Some(e) = de.errors.into_iter().next() {
                return Err(e.into());
            }
            let want = if request.event == AlignmentEvent::Rise {
                EventKind::Rise
            } else {
                EventKind::Set
            };
            for b in de.bodies {
                for e in b.events.into_iter().filter(|e| e.kind == want) {
                    let kind = if want == EventKind::Rise {
                        AlignedKind::Rise
                    } else {
                        AlignedKind::Set
                    };
                    events.push((kind, e.jd_utc, e.az_deg, e.alt_deg));
                }
            }
        }
        AlignmentEvent::AtAltitude { altitude_deg } => {
            if request.options.is_some() {
                return Err(AlmanacError::invalid(
                    "options (the horizon) apply to rise and set only; at_altitude uses the \
                     apparent altitude given",
                ));
            }
            for c in find_altitude(eph, &site, body, w.start, w.end, altitude_deg)? {
                let kind = if c.rising {
                    AlignedKind::Rising
                } else {
                    AlignedKind::Setting
                };
                events.push((kind, c.jd_utc, c.az_deg, c.alt_deg));
            }
        }
    }

    let to_match = |&(kind, jd, az, alt): &(AlignedKind, f64, f64, f64)| AlignmentMatch {
        date: local_date(jd, offset),
        kind,
        jd_utc: jd,
        utc: format_utc(jd),
        az_deg: az,
        offset_deg: norm_180(az - bearing),
        alt_deg: alt,
        best: false,
    };
    let closest = events
        .iter()
        .min_by(|a, b| {
            norm_180(a.2 - bearing)
                .abs()
                .total_cmp(&norm_180(b.2 - bearing).abs())
        })
        .map(to_match);
    let mut matches: Vec<AlignmentMatch> = events
        .iter()
        .filter(|e| norm_180(e.2 - bearing).abs() <= tol)
        .map(to_match)
        .collect();
    mark_best(&mut matches);

    Ok(AlignmentResult {
        body: body.to_string(),
        year: request.year,
        azimuth_deg: bearing,
        tolerance_deg: tol,
        event: request.event,
        utc_offset_hours: offset,
        jd_start: w.start,
        jd_end: w.end,
        truncated: w.truncated,
        events_considered: events.len(),
        matches,
        closest,
    })
}

/// Mark the closest match of each run: matches of one kind less than 1.5 days apart.
fn mark_best(matches: &mut [AlignmentMatch]) {
    let mut done = vec![false; matches.len()];
    for i in 0..matches.len() {
        if done[i] {
            continue;
        }
        let mut run = vec![i];
        let mut last = matches[i].jd_utc;
        for (j, m) in matches.iter().enumerate().skip(i + 1) {
            if done[j] || m.kind != matches[i].kind {
                continue;
            }
            if m.jd_utc - last < 1.5 {
                run.push(j);
                last = m.jd_utc;
            } else {
                break;
            }
        }
        for &j in &run {
            done[j] = true;
        }
        if let Some(&b) = run.iter().min_by(|&&a, &&b| {
            matches[a]
                .offset_deg
                .abs()
                .total_cmp(&matches[b].offset_deg.abs())
        }) {
            matches[b].best = true;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn requests_parse_from_the_wire_shape() {
        let r: AlignmentRequest =
            serde_json::from_str(r#"{"year": 2026, "azimuth_deg": 299, "event": {"kind": "set"}}"#)
                .unwrap();
        assert_eq!(r.body, "Sun");
        assert_eq!(r.tolerance_deg, 0.5);
        assert_eq!(r.event, AlignmentEvent::Set);
        let a: AlignmentRequest = serde_json::from_str(
            r#"{"body": "Moon", "year": 2026, "azimuth_deg": 60, "tolerance_deg": 1,
                "event": {"kind": "at_altitude", "altitude_deg": 5}, "utc_offset_hours": -5}"#,
        )
        .unwrap();
        assert_eq!(a.event, AlignmentEvent::AtAltitude { altitude_deg: 5.0 });
        assert!(
            serde_json::from_str::<AlignmentRequest>(
                r#"{"year": 2026, "azimuth_deg": 299, "event": {"kind": "sunset"}}"#
            )
            .is_err()
        );
    }

    fn m(day: f64, off: f64) -> AlignmentMatch {
        AlignmentMatch {
            date: String::new(),
            kind: AlignedKind::Set,
            jd_utc: 2_461_000.0 + day,
            utc: String::new(),
            az_deg: 299.0 + off,
            offset_deg: off,
            alt_deg: -0.8,
            best: false,
        }
    }

    #[test]
    fn each_run_of_days_has_one_best() {
        let mut v = vec![
            m(0.0, 0.4),
            m(1.0, -0.1),
            m(2.0, -0.45),
            m(45.0, 0.3),
            m(46.0, 0.2),
        ];
        mark_best(&mut v);
        let best: Vec<bool> = v.iter().map(|x| x.best).collect();
        assert_eq!(best, vec![false, true, false, false, true]);
    }
}
